pgjdbc/src/main/java/org/postgresql/util/PGServiceFile.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2018, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.text.MessageFormat;
+import java.util.HashMap;
+
+/*
+ * See https://www.postgresql.org/docs/current/static/libpq-pgservice.html.
+ */
+public class PGServiceFile {
+  public static HashMap load(String service) throws Exception {
+    String filename = findPath();
+    PGServiceFile file = new PGServiceFile();
+    BufferedReader in = new BufferedReader(new FileReader(filename));
+    file.parse(in);
+    return file.getService(service);
+  }
+
+  private static String findPath() {
+    String filename = System.getProperty("postgres.pgservicefile");
+    if (filename == null) {
+      filename = System.getenv().get("PGSERVICEFILE");
+    }
+    if (filename == null) {
+      filename = System.getProperty("user.home") + File.separator + ".pgservices.conf";
+    }
+    return filename;
+  }
+
+  private HashMap<String, HashMap<String, String>> sections;
+
+  public PGServiceFile() {
+    sections = new HashMap();
+  }
+
+  public void parse(BufferedReader in) throws Exception {
+    String line = null;
+    Integer lineno = 0;
+    String section_name = null;
+    HashMap<String, String> section = null;
+
+    while ((line = in.readLine()) != null) {
+      lineno++;
+      line = line.replaceAll("\\s+", "");
```

Comment:
I'm not familiar with the details of INI files but this regex replace would remove whitespace everywhere within the line, not just the ends. Is that intentional? If so I'm curious how they handle passwords with whitespace in them.

Reply:
I took a look at libpq and it's only removing leading whitespace: https://github.com/postgres/postgres/blob/7551d9bc408c2402a8558367ee950ca403e25b37/src/interfaces/libpq/fe-connect.c#L4599

Any whitespace after the start of the keyword is included in the keyword. Ditto for whitesapce in the value, trailing whitespace is included.

Reply:
Thanks for the point. I wasn't sure about this. You're right, `libpq` is more restrictive than Ini4J. Let's match `libpq` behaviour.

Reply:
Done.